File: app/src/main/java/com/cpiekarski/fourteeners/utils/Mountains.java
package com.cpiekarski.fourteeners.utils;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.XmlResourceParser;
import android.util.Log;

import com.cpiekarski.fourteeners.R;

import java.util.ArrayList;
import java.util.List;
import java.util.TreeMap;

/**
 * Maintains a TreeMap for known peaks.
 * 
 * 
 */
public class Mountains {
    
    private final String TAG = "Mountains";
    private TreeMap<String, Mountain> mMnts;
    private Context mCtx;
    // Analytics removed


    private TreeMap<String, ArrayList<String>> mRanges;
    
    public static Mountains instance = null;
    //private TreeMap<Location, String> mLocs;
    
    /**
     * Use Mountains class as a singleton
     * @param context
     * @return
     */
    public static Mountains getInstance(Context context) {
        if(instance == null) {
            instance = new Mountains(context);
            instance.parseFourteeners();
        }
        return instance;
    }
    
    public Mountains(Context context) {
        mMnts = new TreeMap<String, Mountain>();
        mCtx = context;
        mRanges = new TreeMap<String, ArrayList<String>>();

        // no-op: Application subclass removed
    }
    
    public String[] getRanges() {
        return (String[]) mRanges.keySet().toArray(new String[7]);
    }
    
    public String[] getNamesInRange(String range) {
        if ( mRanges.containsKey(range) ) {
            ArrayList<String> al = mRanges.get(range);
            return al.toArray(new String[0]);
        }
        return null;
    }
    
    public String[] getAllPeakNames() {
        return mMnts.keySet().toArray(new String[0]);
    }
    
    /**
     * O(C)
     * @return the number of peaks in the list
     */
    public int getSize() {
        return mMnts.size();
    }
    
    /**
     * Get a known mountain by name only.
     * @param name
     * @return
     */
    public Mountain getMountain(String name) {
        Mountain mnt = mMnts.get(name);
        return mnt;
    }
    
    /**
     * Get a known mountain by name and range.
     * O(log n)
     * 
     * @param name of the Mountain
     * @param range the Mountain is in
     * @return The correct mountain object
     */
    public Mountain getMountain(String name, String range) {
        //TODO: add range check
        Mountain mnt = mMnts.get(name);
        return mnt;
    }
    
    /**
     * O(n log n) add and sort by name
     */
    public void parseFourteeners() {
        SRLOG.v(TAG, "Parsing fourteener data...");
        Resources res = mCtx.getResources();
        XmlResourceParser xrp = res.getXml(R.xml.mountain_data);
        try{
            xrp.next(); // skip first 'mountains' element
            while (xrp.getEventType() != XmlResourceParser.END_DOCUMENT) {
                xrp.next(); // get first 'mountain' element
                if(xrp.getEventType() == XmlResourceParser.START_TAG) {
                    // double check its the right element
                    if(xrp.getName().equals("mountain")) {
                        // extract the data you want
                        int count = xrp.getAttributeCount();
                        String name = xrp.getAttributeValue(null, "name");
                        String rank = xrp.getAttributeValue(null, "rank");
                        String elev = xrp.getAttributeValue(null, "elevation");
                        String range = xrp.getAttributeValue(null, "range");
                        String longitude = xrp.getAttributeValue(null, "long");
                        String latitude = xrp.getAttributeValue(null, "lat");
                        String county = xrp.getAttributeValue(null, "county");

                        Mountain m = new Mountain(name, range , county, Double.parseDouble(longitude), 
                                Double.parseDouble(latitude), Integer.parseInt(rank), Integer.parseInt(elev));
                        mMnts.put(name, m);
                        
                        if(mRanges.containsKey(range)) {
                            ArrayList<String> al = mRanges.get(range);
                            al.add(name);
                        } else {
                            ArrayList<String> nal = new ArrayList<String>();
                            nal.add(name);
                            mRanges.put(range, nal);
                        }
                        
                        SRLOG.v(TAG, "Mountain Attribute Count " + count);
                        SRLOG.v(TAG, "Peak Name " + name);
                        SRLOG.v(TAG, "Peak Elevation " + elev);
                    }
                }
            }
        } catch (Exception e) {
            SRLOG.e(TAG, e.toString());
            // no-op
        } finally {
            xrp.close();
            SRLOG.i(TAG, "Fourteener data parsed");
        }
    }

}

File: app/src/main/kotlin/com/cpiekarski/fourteeners/utils/Mountains.kt
package com.cpiekarski.fourteeners.utils

import android.content.Context
import android.content.res.XmlResourceParser
import com.cpiekarski.fourteeners.R
import com.cpiekarski.fourteeners.SummitRegister
import java.util.*

/**
 * Maintains a TreeMap for known peaks.
 */
class Mountains private constructor(private val context: Context) {
    private val tag = "Mountains"
    private val mountains = TreeMap<String, Mountain>()
    private val ranges = TreeMap<String, ArrayList<String>>()
    private val tracker: Any? // Temporarily Any? instead of Tracker

    companion object {
        @Volatile
        private var INSTANCE: Mountains? = null

        /**
         * Use Mountains class as a singleton
         */
        @JvmStatic
        fun getInstance(context: Context): Mountains {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: Mountains(context.applicationContext).also { 
                    INSTANCE = it
                    it.parseFourteeners()
                }
            }
        }
    }

    init {
        val application = context.applicationContext as SummitRegister
        tracker = application.getDefaultTracker()
    }

    fun getRanges(): Array<String> {
        return ranges.keys.toTypedArray()
    }

    fun getNamesInRange(range: String): Array<String>? {
        return ranges[range]?.toTypedArray()
    }

    fun getAllPeakNames(): Array<String> {
        return mountains.keys.toTypedArray()
    }

    /**
     * O(C)
     * @return the number of peaks in the list
     */
    fun getSize(): Int {
        return mountains.size
    }

    /**
     * Get a known mountain by name only.
     */
    fun getMountain(name: String): Mountain? {
        return mountains[name]
    }

    /**
     * Get a known mountain by name and range.
     * O(log n)
     *
     * @param name of the Mountain
     * @param range the Mountain is in
     * @return The correct mountain object
     */
    fun getMountain(name: String, range: String): Mountain? {
        // TODO: add range check
        return mountains[name]
    }

    /**
     * O(n log n) add and sort by name
     */
    private fun parseFourteeners() {
        SRLOG.v(tag, "Parsing fourteener data...")
        val res = context.resources
        val xrp: XmlResourceParser = res.getXml(R.xml.mountain_data)
        
        try {
            xrp.next() // skip first 'mountains' element
            while (xrp.eventType != XmlResourceParser.END_DOCUMENT) {
                xrp.next() // get first 'mountain' element
                if (xrp.eventType == XmlResourceParser.START_TAG) {
                    // double check it's the right element
                    if (xrp.name == "mountain") {
                        // extract the data you want
                        val count = xrp.attributeCount
                        val name = xrp.getAttributeValue(null, "name")
                        val rank = xrp.getAttributeValue(null, "rank")
                        val elev = xrp.getAttributeValue(null, "elevation")
                        val range = xrp.getAttributeValue(null, "range")
                        val longitude = xrp.getAttributeValue(null, "long")
                        val latitude = xrp.getAttributeValue(null, "lat")
                        val county = xrp.getAttributeValue(null, "county")

                        // Check if all required attributes are present
                        if (name != null && rank != null && elev != null && range != null && 
                            longitude != null && latitude != null && county != null) {
                            
                            val mountain = Mountain(
                                name = name,
                                range = range,
                                county = county,
                                longitude = longitude.toDouble(),
                                latitude = latitude.toDouble(),
                                rank = rank.toInt(),
                                elevation = elev.toInt()
                            )
                            
                            // Only include official fourteeners (rank > 0)
                            if (rank.toInt() > 0) {
                                mountains[name] = mountain
                            }

                            // Only add to ranges if it's an official fourteener
                            if (rank.toInt() > 0) {
                                if (ranges.containsKey(range)) {
                                    ranges[range]?.add(name)
                                } else {
                                    val newList = ArrayList<String>()
                                    newList.add(name)
                                    ranges[range] = newList
                                }
                            }
                        } else {
                            SRLOG.w(tag, "Skipping mountain with missing attributes: name=$name, rank=$rank, elev=$elev, range=$range, longitude=$longitude, latitude=$latitude, county=$county")
                        }

                        SRLOG.v(tag, "Mountain Attribute Count $count")
                        SRLOG.v(tag, "Peak Name $name")
                        SRLOG.v(tag, "Peak Elevation $elev")
                    }
                }
            }
        } catch (e: Exception) {
            SRLOG.e(tag, e.toString())
            // Analytics tracking temporarily commented out
            /*
            tracker?.send(
                HitBuilders.ExceptionBuilder()
                    .setDescription(
                        StandardExceptionParser(context, null)
                            .getDescription(Thread.currentThread().name, e)
                    )
                    .setFatal(false)
                    .build()
            )
            */
        } finally {
            xrp.close()
            SRLOG.i(tag, "Fourteener data parsed")
        }
    }
}